Interface/class method with hyphenated quoted name produces invalid TypeScript

@claude

## Summary

When an interface method signature (or class method) has a quoted string-literal name containing characters that aren't valid in an identifier (e.g. hyphens), Civet strips the quotes in the output, producing invalid TypeScript.

## Reproduction

Input (`.civet`):

```civet
interface Thing
  "do-work"(x: string): string
```

Actual output:

```ts
interface Thing {
  do-work(x: string): string
}
```

Expected output (quotes preserved):

```ts
interface Thing {
  "do-work"(x: string): string
}
```

`tsc` rejects the actual output:

```
error TS1131: Property or signature expected.
error TS1005: ',' expected.
error TS1005: ';' expected.
error TS1128: Declaration or statement expected.
```

The quoted form is valid TypeScript, so Civet should preserve the quotes when the inner content isn't a valid identifier.

## Root cause

`source/parser.hera:4160` in the `MethodSignature` rule strips the surrounding quotes unconditionally whenever the token starts with `"` or `'`:

```civet
name = name.token.match(/^(?:"|')/) ? name.token.slice(1, -1) : name.token
```

It should only strip when the unquoted result is still a valid identifier; otherwise keep the original quoted token.

## Workaround

None other than using a non-hyphenated method name.

--- a/source/parser.hera
+++ b/source/parser.hera
@@ -4157,14 +4157,10 @@ MethodSignature
     if name.name
       name = name.name
     else if name.token
-      // Strip quotes only when the inner content is a valid identifier,
-      // so that hyphenated (or otherwise non-identifier) quoted names
-      // like "do-work" remain quoted in the output (valid TypeScript).
-      if name.token.match(/^(?:"|')/)
-        unquoted := name.token.slice(1, -1)
-        name = unquoted.match(/^[$_\p{ID_Start}][$_\u200C\u200D\p{ID_Continue}]*$/u) ? unquoted : name.token
-      else
-        name = name.token
+      // Preserve quotes as authored in source so that quoted names
+      // (including non-identifier names like "do-work") remain valid
+      // in the output.
+      name = name.token
 
     // TypeScript supports optional methods with bodies; remove ? from JS output
     if optional then optional = { ...optional, ts: true }

--- a/source/parser.hera
+++ b/source/parser.hera
@@ -4152,22 +4152,20 @@ MethodSignature
 
   # NOTE: If this node layout changes, be sure to update
   # `convertMethodToFunction`
-  MethodModifier:modifier ClassElementName:name _?:ws1 QuestionMark?:optional _?:ws2 NonEmptyParameters:parameters ReturnTypeSuffix?:returnType ->
+  MethodModifier:modifier ClassElementName:authoredName _?:ws1 QuestionMark?:optional _?:ws2 NonEmptyParameters:parameters ReturnTypeSuffix?:returnType ->
     // Normalize name so we can check if it is `constructor`
+    name .= authoredName
     if name.name
       name = name.name
     else if name.token
-      // Preserve quotes as authored in source so that quoted names
-      // (including non-identifier names like "do-work") remain valid
-      // in the output.
-      name = name.token
+      name = name.token.match(/^(?:"|')/) ? name.token[<..<] : name.token
 
     // TypeScript supports optional methods with bodies; remove ? from JS output
     if optional then optional = { ...optional, ts: true }
 
     return {
       type: "MethodSignature",
-      children: [ ...modifier.children, name, ws1, optional, ws2, parameters, returnType ],
+      children: [ ...modifier.children, authoredName, ws1, optional, ws2, parameters, returnType ],
       async: modifier.async,
       generator: modifier.generator,
       name,

--- a/source/parser.hera
+++ b/source/parser.hera
@@ -4157,7 +4157,14 @@ MethodSignature
     if name.name
       name = name.name
     else if name.token
-      name = name.token.match(/^(?:"|')/) ? name.token.slice(1, -1) : name.token
+      // Strip quotes only when the inner content is a valid identifier,
+      // so that hyphenated (or otherwise non-identifier) quoted names
+      // like "do-work" remain quoted in the output (valid TypeScript).
+      if name.token.match(/^(?:"|')/)
+        unquoted := name.token.slice(1, -1)
+        name = unquoted.match(/^[$_\p{ID_Start}][$_\u200C\u200D\p{ID_Continue}]*$/u) ? unquoted : name.token
+      else
+        name = name.token
 
     // TypeScript supports optional methods with bodies; remove ? from JS output
     if optional then optional = { ...optional, ts: true }

--- a/test/class.civet
+++ b/test/class.civet
@@ -206,6 +206,34 @@ describe "class", ->
   """
 
   testCase """
+    quoted method with identifier name
+    ---
+    class Thing
+      "foo"(x: string): string
+        return x
+    ---
+    class Thing {
+      foo(x: string): string {
+        return x
+      }
+    }
+  """
+
+  testCase """
+    quoted method with non-identifier name
+    ---
+    class Thing
+      "do-work"(x: string): string
+        return x
+    ---
+    class Thing {
+      "do-work"(x: string): string {
+        return x
+      }
+    }
+  """
+
+  testCase """
     get private method
     ---
     class X {

--- a/test/class.civet
+++ b/test/class.civet
@@ -875,6 +875,24 @@ describe "class", ->
     """
 
     testCase """
+      mixed quoted/unquoted method overloads
+      ---
+      class
+        "f"(x: string): string
+        'f'(x: number): number
+        f(x: unknown): unknown
+          x
+      ---
+      (class {
+        "f"(x: string): string
+        'f'(x: number): number
+        f(x: unknown): unknown {
+          return x
+        }
+      })
+    """
+
+    testCase """
       constructor overloads
       ---
       class

--- a/test/class.civet
+++ b/test/class.civet
@@ -213,7 +213,7 @@ describe "class", ->
         return x
     ---
     class Thing {
-      foo(x: string): string {
+      "foo"(x: string): string {
         return x
       }
     }

--- a/test/types/interface.civet
+++ b/test/types/interface.civet
@@ -109,6 +109,28 @@ describe "[TS] interface", ->
   """
 
   testCase """
+    quoted method with identifier name
+    ---
+    interface Thing
+      "foo"(x: string): string
+    ---
+    interface Thing {
+      foo(x: string): string
+    }
+  """
+
+  testCase """
+    quoted method with non-identifier name
+    ---
+    interface Thing
+      "do-work"(x: string): string
+    ---
+    interface Thing {
+      "do-work"(x: string): string
+    }
+  """
+
+  testCase """
     with get/set
     ---
     interface X {

--- a/test/types/interface.civet
+++ b/test/types/interface.civet
@@ -115,7 +115,7 @@ describe "[TS] interface", ->
       "foo"(x: string): string
     ---
     interface Thing {
-      foo(x: string): string
+      "foo"(x: string): string
     }
   """
 

